src/protocols/secure_channel/CASESession.cpp

```diff
@@ -404,187 +445,283 @@ CHIP_ERROR CASESession::HandleSigmaR1(System::PacketBufferHandle && msg)
+                                         msg_R2_Encrypted.Get() + msg_r2_signed_enc_len, kTAGSize));
 
     // Construct Sigma2 Msg
-    data_len = EstimateTLVStructOverhead(
+    size_t data_len = EstimateTLVStructOverhead(
         kSigmaParamRandomNumberSize + sizeof(uint16_t) + kP256_PublicKey_Length + msg_r2_signed_enc_len + kTAGSize, 4);
 
-    msg_R2 = System::PacketBufferHandle::New(data_len);
-    VerifyOrExit(!msg_R2.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+    System::PacketBufferHandle msg_R2 = System::PacketBufferHandle::New(data_len);
+    VerifyOrReturnError(!msg_R2.IsNull(), CHIP_ERROR_NO_MEMORY);
 
-    {
-        System::PacketBufferTLVWriter tlvWriter;
-        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    System::PacketBufferTLVWriter tlvWriterMsg2;
+    outerContainerType = TLV::kTLVType_NotSpecified;
 
-        tlvWriter.Init(std::move(msg_R2));
-        SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
-        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(1), &msg_rand[0], sizeof(msg_rand)));
-        SuccessOrExit(err = tlvWriter.Put(TLV::ContextTag(2), GetLocalSessionId(), true));
-        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(3), mEphemeralKey.Pubkey(),
-                                               static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
-        SuccessOrExit(err = tlvWriter.PutBytes(TLV::ContextTag(4), msg_R2_Encrypted.Get(),
-                                               static_cast<uint32_t>(msg_r2_signed_enc_len + kTAGSize)));
-        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
-        SuccessOrExit(err = tlvWriter.Finalize(&msg_R2));
-    }
+    tlvWriterMsg2.Init(std::move(msg_R2));
+    ReturnErrorOnFailure(tlvWriterMsg2.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(1), &msg_rand[0], sizeof(msg_rand)));
+    ReturnErrorOnFailure(tlvWriterMsg2.Put(TLV::ContextTag(2), GetLocalSessionId(), true));
+    ReturnErrorOnFailure(
+        tlvWriterMsg2.PutBytes(TLV::ContextTag(3), mEphemeralKey.Pubkey(), static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
+    ReturnErrorOnFailure(tlvWriterMsg2.PutBytes(TLV::ContextTag(4), msg_R2_Encrypted.Get(),
+                                                static_cast<uint32_t>(msg_r2_signed_enc_len + kTAGSize)));
+    ReturnErrorOnFailure(tlvWriterMsg2.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriterMsg2.Finalize(&msg_R2));
 
-    err = mCommissioningHash.AddData(ByteSpan{ msg_R2->Start(), msg_R2->DataLength() });
-    SuccessOrExit(err);
+    ReturnErrorOnFailure(mCommissioningHash.AddData(ByteSpan{ msg_R2->Start(), msg_R2->DataLength() }));
 
-    mNextExpectedMsg = Protocols::SecureChannel::MsgType::CASE_SigmaR3;
+    mState = kSentSigma2;
```

Comment:
This should be moved after the SendMessage

Reply:
https://github.com/project-chip/connectedhomeip/pull/9904#discussion_r715178972